motr/client.c

```diff
@@ -830,84 +831,116 @@ M0_INTERNAL int m0_op_init(struct m0_op *op,
 	return M0_RC(0);
 }
 
-M0_INTERNAL int m0_calculate_md5_inc_digest(
-		struct m0_md5_inc_digest_pi *pi,
+#ifndef __KERNEL__
+M0_INTERNAL int m0_calculate_md5_inc_context(
+		struct m0_md5_inc_context_pi *pi,
 		struct m0_pi_seed *seed,
 		struct m0_bufvec *bvec,
 		enum m0_pi_calc_flag flag,
-		unsigned char *curr_digest,
+		unsigned char *curr_context,
 		unsigned char *pi_value_without_seed)
 {
+	MD5_CTX context;
+	int i, rc;
+
 	M0_ENTRY();
 
 	M0_PRE(pi != NULL);
-	M0_PRE(curr_digest != NULL);
+	M0_PRE(curr_context != NULL);
 	M0_PRE(bvec != NULL);
 	M0_PRE(bvec->ov_vec.v_count != NULL);
 	M0_PRE(bvec->ov_buf != NULL);
 
-	/* Following code block should be removed */
-	/* sending dummy values to S3 */
-#if 1
-#define PI_DIGEST_LENGTH 92
-#define PI_VALUE_LENGTH 16
+	/* This call is for first data unit, need to initialize prev_context */
 	if (flag & M0_PI_CALC_UNIT_ZERO) {
-		 memset(&pi->prev_digest, '0', PI_DIGEST_LENGTH);
-	}
-	memset(&pi->pi_value, 'A', PI_VALUE_LENGTH);
-	memset(curr_digest, '1', PI_DIGEST_LENGTH);
-	if (pi_value_without_seed) {
-		memset(pi_value_without_seed, 'B', PI_VALUE_LENGTH);
+		rc = MD5_Init((MD5_CTX *)&pi->prev_context);
+		if (rc != 1) {
+			return M0_ERR_INFO(rc, "MD5_Init failed.");
+		}
 	}
-#endif
 
-	/* Uncomment following code once motr has integration for MD5 */
-#if 0
+	/* memcpy, so that we do not change the prev_context */
+	memcpy(curr_context, &pi->prev_context, sizeof(MD5_CTX));
```

Comment:
Is this correct 
memcpy(curr_context, &pi->prev_context, sizeof(MD5_CTX));
or should it be
memcpy(curr_context, pi->prev_context, sizeof(MD5_CTX));

prev_context is the base address of the array of size MD5_CTX